tracing-subscriber/src/layer/mod.rs

```diff
@@ -1160,13 +1164,31 @@ where
     L: Layer<S>,
     S: Subscriber,
 {
+    fn on_layer(&mut self, subscriber: &mut S) {
+        if let Some(inner) = Arc::get_mut(self) {
+            // XXX(eliza): this may behave weird if another `Arc` clone of this
+            // layer is layered onto a _different_ subscriber...but there's no
+            // good solution for that...
```

Comment:
I don't understand this comment. `Arc::get_mut` will only return a value if there are no other aliases so if it gets here there *can't* be a different subscriber. So the weird behaviour will be if there are multiple references to the Layer then this won't get called.

It seems like `Arc<L: Layer>` and `on_layer` are intrinsically incompatible. If having `on_layer(&mut self, ...` is useful, then maybe there shouldn't be Arc<L: Layer> after all?

Reply:
> `Arc::get_mut` will only return a value if there are no other aliases so if it gets here there _can't_ be a different subscriber.

 was thinking about a situation where the `Arc` is cloned _after_ `on_layer` is called. But, in practice, that will never happen, because `on_layer` is only called when layering, and layering moves the layer by value. However, there's nothing _stopping_ user code from just arbitrarily calling `on_layer`.

> It seems like `Arc<Layer>` and `on_layer` are intrinsically incompatible. If having `on_layer(&mut self, ...` is useful, then maybe there shouldn't be Arc after all?

Yes, I'm thinking that as well --- I regret not thinking through the potential implications of the `Arc` impl w.r.t having an `on_layer` callback. We may want to deprecate the `Arc` impls...which doesn't feel great, since we _just_ added them...

Reply:
...ugh, i've been once again reminded that you can't deprecate a trait impl for a type: 
![image](https://user-images.githubusercontent.com/2796466/133895750-53bbc417-f3aa-4ceb-8e07-f73d58d3b20f.png)

so, I guess the only real solution for this is a big warning in the docs and removing it in the next breaking change. Gah!

Reply:
Better to deprecate a bad idea earlier than later. And if Arc<Layer> isn't actually usable, then presumably that  will dissuade people from using it.